packages/discovery-provider/plugins/pedalboard/apps/backfill-audio-analyses/src/backfill.ts

```diff
@@ -0,0 +1,164 @@
+import axios from 'axios'
+import _ from 'lodash'
+import { ethers } from 'ethers'
+import { Semaphore } from 'await-semaphore'
+import { Knex } from 'knex'
+import { App } from '@pedalboard/basekit'
+import { config } from '.'
+import { SharedData } from './index'
+import {
+  storeDbOffset,
+  readDbOffset,
+  getCachedHealthyContentNodes
+} from './redis'
+
+// Batch size for fetching tracks
+const BACKFILL_BATCH_SIZE = config.testRun ? 100 : 1000
+
+// Concurrency control
+const MAX_CONCURRENT_REQUESTS = 10
+const semaphore = new Semaphore(MAX_CONCURRENT_REQUESTS)
+
+const REQUEST_TIMEOUT = 30000 // 30s
+
+interface Track {
+  track_id: number
+  audio_upload_id: string
+  musical_key: string | null
+  bpm: number | null
+  audio_analysis_error_count: number | null
+}
+
+function generateSignature(data: any, privateKey: string): string {
+  const toSignStr = JSON.stringify(data, Object.keys(data).sort())
+  const toSignHash = ethers.utils.keccak256(ethers.utils.toUtf8Bytes(toSignStr))
+  const signingKey = new ethers.utils.SigningKey(privateKey)
+  const signature = signingKey.signDigest(toSignHash)
+  return ethers.utils.joinSignature(signature)
+}
+
+async function analyzeAudio(
+  contentNodes: string[],
+  track: Track
+): Promise<void> {
+  const audioUploadId = track.audio_upload_id
+
+  const release = await semaphore.acquire() // Acquire a semaphore permit
```

Comment:
n00b question, what is this semaphore for? if we're running this from a [for loop](https://github.com/AudiusProject/audius-protocol/pull/8788/files#diff-4058f6191df6964b25bf84f904ef23291fb16f8c722ce3d30933fd0b24d269b2R110) and waiting 30s is there anything to synchronize?

Reply:
its so it can run 10 `analyzeAudio`s at a time as it works through each batch - see the `await Promise.all(analyzePromises)` on line 134 below